cpu/stm32_common/periph/uart.c

```diff
@@ -43,19 +43,57 @@ static inline USART_TypeDef *dev(uart_t uart)
     return uart_config[uart].dev;
 }
 
-int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
-{
+int uart_set_baudrate(uart_t uart, uint32_t baudrate) {
     uint16_t mantissa;
     uint8_t fraction;
-    uint32_t clk;
+    uint32_t uart_clk;
+
+    uart_clk = periph_apb_clk(uart_config[uart].bus)/baudrate;
+
+    if (uart_clk < 8) {
+        /* clock is too slow for using UART with specified baudrate */
+        periph_clk_dis(uart_config[uart].bus, uart_config[uart].rcc_mask);
+        return UART_NOBAUD;
+    } else {
+        /* Disable UART. Setting BRR on enabled USART1 sometimes somehow results in Hard Fault */
+        dev(uart)->CR1 &= ~USART_CR1_UE;
+
+        /* choose between 8x and 16x oversampling */
+        /* 16x is preferred, but is not possible on low clock frequency */
+#ifdef USART_CR1_OVER8
+        if (uart_clk < 16) {
+            dev(uart)->CR1 |= USART_CR1_OVER8;
+            mantissa = (uint16_t)(uart_clk / 8);
+            fraction = (uint8_t)(uart_clk - (mantissa * 8));
+            dev(uart)->BRR = ((mantissa & 0x0fff) << 4) | (fraction & 0x07);
+        } else {
+            dev(uart)->CR1 &= ~USART_CR1_OVER8;
+            mantissa = (uint16_t)(uart_clk / 16);
+            fraction = (uint8_t)(uart_clk - (mantissa * 16));
+            dev(uart)->BRR = ((mantissa & 0x0fff) << 4) | (fraction & 0x0f);
+        }
+#else
+        mantissa = (uint16_t)(uart_clk / 16);
+        fraction = (uint8_t)(uart_clk - (mantissa * 16));
+        dev(uart)->BRR = ((mantissa & 0x0fff) << 4) | (fraction & 0x0f);
+#endif
```

Comment:
Can you explain all the magic numbers in this function? The hex are understandable but I'd prefer to see the 8 and 16 as constants like `#define UART_8X_OVERSAMPLING   8`, or something like that.